xds/src/test/java/io/grpc/xds/internal/certprovider/FileWatcherCertificateProviderTest.java

```diff
@@ -0,0 +1,343 @@
+      boolean deleteCurRoot)
+      throws IOException {
+    if (deleteCurCert) {
+      Files.delete(Paths.get(certFile));
+    }
+    if (certFileSource != null) {
+      certFileSource = CommonTlsContextTestsUtil.getTempFileNameForResourcesFile(certFileSource);
+      Files.copy(Paths.get(certFileSource), Paths.get(certFile), REPLACE_EXISTING);
+    }
+    if (deleteCurKey) {
+      Files.delete(Paths.get(keyFile));
+    }
+    if (keyFileSource != null) {
+      keyFileSource = CommonTlsContextTestsUtil.getTempFileNameForResourcesFile(keyFileSource);
+      Files.copy(Paths.get(keyFileSource), Paths.get(keyFile), REPLACE_EXISTING);
+    }
+    if (deleteCurRoot) {
+      Files.delete(Paths.get(rootFile));
+    }
+    if (rootFileSource != null) {
+      rootFileSource = CommonTlsContextTestsUtil.getTempFileNameForResourcesFile(rootFileSource);
+      Files.copy(Paths.get(rootFileSource), Paths.get(rootFile), REPLACE_EXISTING);
+    }
+  }
+
+  @Test
+  public void getCertificateAndCheckUpdates()
+      throws IOException, CertificateException, InterruptedException {
+    MeshCaCertificateProviderTest.TestScheduledFuture<?> scheduledFuture =
+        new MeshCaCertificateProviderTest.TestScheduledFuture<>();
+    doReturn(scheduledFuture)
+        .when(timeService)
+        .schedule(any(Runnable.class), any(Long.TYPE), eq(TimeUnit.SECONDS));
+    populateTarget(CLIENT_PEM_FILE, CLIENT_KEY_FILE, CA_PEM_FILE, false, false, false);
+    provider.checkAndReloadCertificates();
+    verifyWatcherUpdates(CLIENT_PEM_FILE, CA_PEM_FILE);
+    verifyTimeServiceAndScheduledHandle();
+
+    reset(mockWatcher, timeService);
+    doReturn(scheduledFuture)
+        .when(timeService)
+        .schedule(any(Runnable.class), any(Long.TYPE), eq(TimeUnit.SECONDS));
+    provider.checkAndReloadCertificates();
+    verifyWatcherErrorUpdates(null, null, 0, 0, (String[]) null);
+    verifyTimeServiceAndScheduledHandle();
+
+    reset(mockWatcher, timeService);
+    doReturn(scheduledFuture)
+        .when(timeService)
+        .schedule(any(Runnable.class), any(Long.TYPE), eq(TimeUnit.SECONDS));
+    Thread.sleep(1000L);
+    populateTarget(null, null, SERVER_1_PEM_FILE, false, false, false);
+    provider.checkAndReloadCertificates();
+    verifyWatcherUpdates(null, SERVER_1_PEM_FILE);
+    verifyTimeServiceAndScheduledHandle();
+
+    reset(mockWatcher, timeService);
+    doReturn(scheduledFuture)
+        .when(timeService)
+        .schedule(any(Runnable.class), any(Long.TYPE), eq(TimeUnit.SECONDS));
+    Thread.sleep(1000L);
+    populateTarget(SERVER_0_PEM_FILE, SERVER_0_KEY_FILE, null, false, false, false);
+    provider.checkAndReloadCertificates();
+    verifyWatcherUpdates(SERVER_0_PEM_FILE, null);
+    verifyTimeServiceAndScheduledHandle();
+
+    reset(mockWatcher, timeService);
+    doReturn(scheduledFuture)
+        .when(timeService)
+        .schedule(any(Runnable.class), any(Long.TYPE), eq(TimeUnit.SECONDS));
+    Thread.sleep(1000L);
```

Comment:
The test might take a little too long, 5s. Is it possible to make use of FakeClock for scheduler like other tests are doing?

Reply:
The problem is the file system timestamp precision is 1 second and the test uses actual files with their time stamps which cannot be faked. Hence 1 sec delay for unequal time stamps

Reply:
I see. Then can we separate them in multiple tests, each sleeps shorter? So if we run tests in parallel with gradle, it can save some time.

Reply:
Good idea. Done.